internal/querynodev2/tasks/concurrent_safe_scheduler.go

```diff
@@ -0,0 +1,240 @@
+package tasks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/milvus-io/milvus/internal/querynodev2/collector"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/metrics"
+	"github.com/milvus-io/milvus/pkg/util/conc"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+	"github.com/milvus-io/milvus/pkg/util/metricsinfo"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+const (
+	maxReceiveChanBatchConsumeNum = 100
+)
+
+func newConcurrentSafeScheduler(policy schedulePolicy) Scheduler {
+	maxReadConcurrency := paramtable.Get().QueryNodeCfg.MaxReadConcurrency.GetAsInt()
+	maxReceiveChanSize := paramtable.Get().QueryNodeCfg.MaxReceiveChanSize.GetAsInt()
+	log.Info("query node use concurrent safe scheduler", zap.Int("max_concurrency", maxReadConcurrency))
+	return &concurrentSafeScheduler{
+		policy:           policy,
+		receiveChan:      make(chan addTaskReq, maxReceiveChanSize),
+		execChan:         make(chan Task),
+		pool:             conc.NewPool[any](maxReadConcurrency, conc.WithPreAlloc(true)),
+		schedulerCounter: schedulerCounter{},
+	}
+}
+
+type addTaskReq struct {
+	task Task
+	err  chan<- error
+}
+
+// concurrentSafeScheduler is a general concurrent safe scheduler implementation by wrapping a schedule policy.
+type concurrentSafeScheduler struct {
```

Comment:
simply name it scheduler (or any nice name), the concurrent safe property is not important for this as it's not a data structure, and have to be always safe in concurrency env